tests/test_download.py

```diff
@@ -0,0 +1,114 @@
+import os
+import shutil
+import tempfile
+import hashlib
+import zipfile
+from pathlib import Path
+import unittest
+from unittest import mock
+
+
+from sister import download
+
+
+class TestSetCacheRoot(unittest.TestCase):
+
+    def test_set_cache_root(self):
+        orig_root = download.get_cache_root()
+        new_root = '/tmp/cache'
+        try:
+            download.set_cache_root(new_root)
+            self.assertEqual(download.get_cache_root(), new_root)
+        finally:
+            download.set_cache_root(orig_root)
+
+
+class TestGetCacheDirectory(unittest.TestCase):
+
+    def test_get_cache_directory(self):
+        root = download.get_cache_root()
+        path = download.get_cache_directory('test', False)
+        self.assertEqual(path, os.path.join(root, 'test'))
+
+    def test_fails_to_make_directory(self):
+        with mock.patch('os.makedirs') as f:
+            f.side_effect = OSError()
+            with self.assertRaises(OSError):
+                download.get_cache_directory('/sister_test_cache', True)
+
+
+class TestCachedDownload(unittest.TestCase):
+
+    def setUp(self):
+        self.default_cache_root = download.get_cache_root()
+        self.temp_dir = tempfile.mkdtemp()
+        download.set_cache_root(self.temp_dir)
+
+    def tearDown(self):
+        download.set_cache_root(self.default_cache_root)
+        shutil.rmtree(self.temp_dir)
+
+    def test_fails_to_make_directory(self):
+        with mock.patch('os.makedirs') as f:
+            f.side_effect = OSError()
+            with self.assertRaises(OSError):
+                download.cached_download('https://example.com')
+
+    def test_file_exists(self):
+        # Make an empty file which has the same name as the cache directory
+        with open(os.path.join(self.temp_dir, '_dl_cache'), 'w'):
+            pass
+        with self.assertRaises(OSError):
+            download.cached_download('https://example.com')
+
+    def test_cache_exists(self):
+        with mock.patch('os.path.exists') as f:
+            f.return_value = True
+            url = 'https://example.com'
+            path = download.cached_download(url)
+            self.assertEqual(path, f'{self.temp_dir}/_dl_cache/{hashlib.md5(url.encode("utf-8")).hexdigest()}')
```

Comment:
How about using decorator rather than context manager like below?:
```py
@mock.patch('os.path.exists')
def test_cache_exists(self, f):
    f.return_value = True
    url = 'https://example.com'
    path = donwload.cached_download(url)
    self.assertEqual(path, os.path.join(self.temp_dir, '_dl_cache', hashlib.md5(url.encode('utf-8')).hexdigest()))
```

Reply:
A test function taking an argument (`f`) appeals a bit weird to me.
Can you share why you think using a decorator might fit here?

Reply:
I guess the flat code is more readable than the nested code. 
But this isn't a big deal. So you can just ignore it if you feel weird.

Reply:
Also, some might think the decorator version is hard to understand.

Reply:
> flat code is more readable than the nested code
Good point.

As you said it’s not a big deal.
So for now I will go with this decorator way (which is new to me) and if I still feel strange some time later,
I will change back to the nested one.


 Thanks

Reply:
I said "It's not a big deal.", because I thought the context manager way is not very nested, 
So it's up to you. I respect your decision 😉